GVFS/GVFS.Common/Cleanup/GitCleanupQueue.cs

```diff
@@ -0,0 +1,155 @@
+using System.IO;
+using System.Threading;
+
+namespace GVFS.Common.Cleanup
+{
+    public class GitCleanupQueue
+    {
+        private readonly object threadLock = new object();
+        private readonly object currentStepLock = new object();
+        private GVFSContext context;
+        private ConcurrentQueue<GitCleanupStep> queue;
+        private Thread thread;
+        private GitCleanupStep currentStep;
+        private bool stopping;
+
+        public GitCleanupQueue(GVFSContext context)
+        {
+            this.context = context;
+            this.queue = new ConcurrentQueue<GitCleanupStep>();
+        }
+
+        public void Enqueue(GitCleanupStep step)
+        {
+            this.queue.Enqueue(step);
+
+            lock (this.threadLock)
+            {
+                if (this.thread == null)
+                {
+                    this.thread = new Thread(() => this.RunQueue());
+                    this.thread.IsBackground = true;
+
+                    try
+                    {
+                        this.thread.Start();
+                    }
+                    catch (ThreadStateException e)
+                    {
+                        this.LogError(nameof(GitCleanupQueue), nameof(this.Enqueue), e);
+                    }
+                    catch (OutOfMemoryException e)
+                    {
+                        this.LogError(nameof(GitCleanupQueue), nameof(this.Enqueue), e);
+                    }
+                }
+            }
+        }
+
+        public void Stop()
+        {
+            this.stopping = true;
+
+            GitCleanupStep stepToStop;
+
+            lock (this.currentStepLock)
+            {
+                stepToStop = this.currentStep;
+            }
+
+            if (stepToStop != null)
+            {
+                stepToStop.Stop();
+            }
+        }
+
+        /// <summary>
+        /// This method is used for test purposes only.
+        /// </summary>
+        public void WaitForStepsToFinish()
+        {
+            this.thread?.Join();
+        }
+
+        /// <summary>
+        /// This method is public for test purposes only.
+        /// </summary>
+        public bool EnlistmentRootReady()
+        {
+            // If a user locks their drive or disconnects an external drive while the mount process
+            // is running, then it will appear as if the directories below do not exist or throw
+            // a "Device is not ready" error.
+            try
+            {
+                return this.context.FileSystem.DirectoryExists(this.context.Enlistment.EnlistmentRoot)
+                         && this.context.FileSystem.DirectoryExists(this.context.Enlistment.GitObjectsRoot);
+            }
+            catch (IOException)
+            {
+                return false;
+            }
+        }
+
+        private void RunQueue()
+        {
+            while (true)
+            {
+                lock (this.threadLock)

```

Comment:
Can the lock just be around the:
this.thread = null;
portion?

There is also a very small race condition here where the drive becomes unavailable after the EnlistmentRootReady check, but the risk there is pretty small and we really don't have another option.

Reply:
But back to your actual question, I do think the lock could just be around the assignment.